lib/Optimizer/GraphOptimizer/ConstantFolding.cpp

```
@@ -0,0 +1,276 @@
+  auto *clonedC = recursiveClone(constEvaluationF, C, currToNew);
+  // Create save nodes for each of the results.
+  llvm::SmallVector<SaveNode *, 16> savedResults;
+  for (unsigned idx = 0, e = clonedC->getNumResults(); idx < e; ++idx) {
+    auto *SN = constEvaluationF->createSave(clonedC->getName(),
+                                            clonedC->getNthResult(idx));
+    savedResults.emplace_back(SN);
+    bindings.allocate(SN->getPlaceholder());
+  }
+  // Run the temporary backend to perform this constant operation
+  // evaluation.
+  EXIT_ON_ERR(
+      executeConstantFunction(backend, *constEvaluationF, bindings, cctx));
+  // Get the results of the constant operation compile-time computation and
+  // create new constants from it.
+  std::vector<Constant *> constResults;
+  constResults.reserve(savedResults.size());
+  for (auto *SN : savedResults) {
+    Tensor *outputTensor = bindings.get(SN->getPlaceholder());
+    auto *constResult = mod.createConstant(SN->getName(), *outputTensor);
+    constResults.emplace_back(constResult);
+  }
+  // Remove the temporary function.
+  mod.eraseFunction(constEvaluationF);
+  return constResults;
+}
+
+/// Check if function \p F consists of constant operations only.
+llvm::Error verifyConstantFunction(Backend &backend, Function &F) {
+  // Perform the checks in DEBUG builds only.
+#ifndef NDEBUG
+  for (auto &N : F.getNodes()) {
+    // Saving results is fine.
+    if (isa<SaveNode>(&N)) {
+      continue;
+    }
+    // Placeholders can be used just to save results.
+    if (!isa<Placeholder>(&N)) {
+      RETURN_ERR_IF_NOT(isConstantOperation(&N, backend),
+                        "Expected constant operation");
+      continue;
+    }
+    if (!N.hasOneUse()) {
+      RETURN_ERR("Expected constant operation");
+    }
+    auto *SN = dyn_cast<SaveNode>(N.getUsers().begin()->getUser());
+    if (SN && SN->getPlaceholder() == &N) {
+      continue;
+    }
```

Don't you need to return an error at the end of the for loop here, since it's a PH that isn't used by a Save?

---

Good catch! Thanks!